Raw parser

JFRudzinski opened the pull request:
# Add Bruker/Siemens RAW v4 Format Support

## Summary
Adds support for parsing Bruker/Siemens RAW v4 binary XRD files with automatic X-ray wavelength detection via anode material lookup.

## New Files
- **`src/fairmat_readers_xrd/bruker_raw_parser.py`** - Binary parser with wavelength reference table for 6 anode materials (Cu, Mo, Co, Fe, Cr, Ag)
- **`notes/RAW_FORMAT_NOTES.md`** - Technical documentation of binary format structure and offset mappings
- **`tests/test_bruker_raw.py`** - Comprehensive unit tests for parser functionality
- **`tests/data/TwoTheta_scan_scrambled.raw`** - Anonymized test file (scrambled intensities protect client data)

## Modified Files
- **`src/fairmat_readers_xrd/readers.py`** - Added `read_bruker_raw()` function with source metadata extraction
- **`tests/test_readers.py`** - Added RAW format tests validating structure, units, and metadata
- **`README.md`** - Updated format support list and API documentation

## Key Features
- **Automatic Wavelength Detection**: Reads anode material at offset `0x01A8`, looks up K-alpha1/K-alpha2/K-beta wavelengths from International Tables for Crystallography
- **Complete Metadata Extraction**: Scan parameters, measurement conditions, 16 measurement ranges
- **Paired File Support**: Optionally reads paired XRDML for additional metadata
- **Data Protection**: Test files use scrambled intensities while preserving binary structure

JFRudzinski committed:
basic rigaku raw parsing

JFRudzinski committed:
fix shapes

JFRudzinski committed:
adjustments and implementation notes

JFRudzinski committed:
ruff

JFRudzinski committed:
ruff again

JFRudzinski commented:
@ka-sarthak I just ended up implementing a simple parser for Rigaku `.raw`. Actually, beyond the difficulties of running the 3rd party windows software, I found some strange behavior of the powdll converter. I left some notes in `RAW_FORMAT_NOTES.md` for you about this, maybe this is interesting to save somewhere for future reference.

Please have a look at both PRs when you have a chance. :pray:

ka-sarthak commented on notes/RAW_FORMAT_NOTES.md:
```diff
@@ -0,0 +1,237 @@
+
+## Validation Results
+
+Tested on sample file: `HeOx-1001-nsp-sps-900C-10min-01-poliert_exported.raw`
```
Can you add the raw file in the test folder and use it in the `test_rigaku_raw.py`?

ka-sarthak commented on src/fairmat_readers_xrd/readers.py:
```diff
@@ -301,6 +308,76 @@ def read_bruker_brml(file_path: str, logger: 'BoundLogger' = None) -> Dict[str,
+        metadata = {
+            'sample_id': raw_data['metadata'].get('sample_id'),
+            'scan_type': scan_type,
+            'scan_axis': '2Theta',
```
Since `scan_type` is not available in the raw file, let's not populate it with an expected value (which might not be true if scan axis is Phi or Chi or Omega). Instead, let's leave this key-val empty.

ka-sarthak commented on src/fairmat_readers_xrd/readers.py:
```diff
@@ -301,6 +308,76 @@ def read_bruker_brml(file_path: str, logger: 'BoundLogger' = None) -> Dict[str,
+        # Return data in the same format as read_panalytical_xrdml
+        return {
+            **modified_scan_data,
+            'scanmotname': '2Theta',
```
Since `scanmotname` is not available in the raw file, let's not populate it with an expected value. Instead, let's leave this key-val empty.

ka-sarthak commented on src/fairmat_readers_xrd/readers.py:
Please add another if condition for returning the `read_rigaku_raw` if file path ends with `.raw` .

ka-sarthak commented on tests/test_readers.py:
```diff
@@ -87,3 +90,57 @@ def test_brml_reader():
+    # Note: This test requires a sample .raw file to be added to tests/data/
+    # For now, we test the basic functionality without a reference JSON
+    # Check if a test RAW file exists
+    test_raw = 'tests/data/test_sample.raw'
```
Can we add a test file? You can scramble the intensity two-theta data and anonymize the metadata if required.

ka-sarthak commented on tests/generate_test_reference.py:
```diff
@@ -0,0 +1,118 @@
+    return obj
+
+
+def generate_reference(raw_file: Path, output_file: Path = None):
```
What is the purpose of this function? Is it to create a reference JSON simulating the output of the reader?

ka-sarthak reviewed (changes requested):
It looks great. I was wondering if the `.raw` file includes the data of other axis positions like phi, chi, omega. We are extracting these values from Rigaku `.rasx` files. Can you check if that's case if it's not too much effort?

Else we can assume that the angles always point to two_theta values. I added some comments to leave the scan_axis and scanmotname as empty since this information is not available. The nomad-measurement XRD schemas expect that this information might be missing, so we can avoid making the assumptions while reading.

Lastly, I see you have added a lot of extensive testing for the reader function and class. It will be super nice to add a test datafile with dummy data along with it.

JFRudzinski committed:
adjustments from code review, Rigaku to Bruker in code, warnings to clarify scope

JFRudzinski committed:
module and function renames rigaku to bruker

JFRudzinski committed:
get source metadata from table

JFRudzinski committed:
removal of sample_id setting in reader

JFRudzinski committed:
updated README

JFRudzinski committed:
ruff

JFRudzinski committed:
ruff update

JFRudzinski committed:
update tests

JFRudzinski committed:
remove data generation

JFRudzinski committed:
ruff format

JFRudzinski commented on notes/RAW_FORMAT_NOTES.md:
```diff
@@ -0,0 +1,310 @@
+# Siemens/Bruker RAW v4 Format - Reverse Engineering Documentation
```
@ka-sarthak I would remove this before merging unless you want to keep it somewhere

ka-sarthak replied:
I think it's fine you wanna keep it here under notes. I would suggest moving `TESTING_RAW_FILES.md` to the `notes/` folder.

JFRudzinski commented on src/fairmat_readers_xrd/bruker_raw_parser.py:
```diff
@@ -0,0 +1,409 @@
+
+# X-ray wavelength reference table (in Angstroms)
+# Source: International Tables for Crystallography, Volume C
+XRAY_WAVELENGTHS = {
```
@ka-sarthak what do you make of this table look up approach? I could only parse the source element and the model seems to think this is what powdll is doing to convert from raw to xrdml

ka-sarthak replied:
I like this approach. Found this reference. We can add this as a comment: https://onlinelibrary.wiley.com/iucr/itc/Cb/ch4o2v0001/sec4o2o2/

The values are slightly different than what we have here. Can you revise them, keeping 6 decimal point precision?

ka-sarthak commented on src/fairmat_readers_xrd/readers.py:
```diff
@@ -355,24 +359,41 @@ def read_rigaku_raw(
+
         metadata = {
-            'sample_id': raw_data['metadata'].get('sample_id'),
+            'sample_id': None,  # Not extracted for consistency with other formats
```
I think we are extracting sample_id in at least RASX format. If this information is available in the raw file, we can add it here.

JFRudzinski replied:
ok, I originally resolved this convo, but now I am a little unsure: 

I did re-instantiate the extraction of sample_id from the raw file as I previously had. I do not see however that rasx or any other format is extracting the sample_id. 

I guess this doesn't matter, but what I see now is that when I set the sample_id, this is stored in `data.samples[0].lab_id`, but this is not carried over to data.lab_id which is shown as empty on the overview page. Is this intended / normal usage of the ELN? Or in case there is only 1 sample in the upload should the lab_id from samples[0] be carried over to the global lab_id?

ka-sarthak replied:
Sample ID from the rawfile will be used in the [measurement ELN](https://github.com/FAIRmat-NFDI/nomad-measurements/blob/main/src/nomad_measurements/xrd/schema.py#L1784) for automatically connecting it with another entry (sample entry) that has the same sample ID stored in the 'results.eln.lab_ids' quantity. This functionality comes from the base section `CompositeSystemReference(EntityReference)`.

The lab ID of the measurement ELN is, however, assumed to be different than that of the sample. Therefore, we do not populate `data.lab_id` with `data.samples[0].lab_id`. This assumption can be false for specific labs, in which case the ELN can be extended to have an additional normalization step for this.

For readers in this repository, extracting the sample ID, if available, is a good idea.

ka-sarthak commented on tests/generate_test_reference.py:
```diff
@@ -46,7 +46,15 @@ def convert_quantity_to_string(obj):
 
 def generate_reference(raw_file: Path, output_file: Path = None):
```
We are using the `convert_quantity_to_string` function under `tests/test_readers.py` for almost the same purpose. It reduces the reader's output to strings, and the keys containing arrays are modified to contain only the shape of the arrays. This string is then compared to the string created from a reference.

The reference is loaded from a JSON saved under `tests/data`. I created this reference once while adding the tests using:
```py
reference_str = convert_quantity_to_string(reader_func(raw_file))
with open('path_to_reference.json', 'w') as f:
    json.dump(json.loads(reference_str), f, indent=2)
```

Maybe `convert_quantity_to_string` for building this test too.

JFRudzinski replied:
no longer relevant, removed this test file in favor of standard tests

ka-sarthak commented on tests/test_bruker_raw.py:
```diff
@@ -57,7 +57,7 @@ def sample_raw_file(self):
     def test_parser_initialization(self):
         """Test that parser can be initialized with a file path."""
-        parser = RigakuRAW4Parser('dummy.raw')
+        parser = BrukerRAW4Parser('dummy.raw')
```
"dummy.raw" can be replaced with "data/test_sample.raw"?

JFRudzinski replied:
This test does not actually depend on a file being present, so I think it is somewhat misleading to add a path to real test data, or do you see it differently? Am I missing something?

ka-sarthak replied:
Ah, I see. Then it looks fine!

ka-sarthak commented on tests/TESTING_RAW_FILES.md:
```diff
@@ -0,0 +1,202 @@
+# Testing Rigaku RAW File Parser
```
This can be moved to `notes/` folder

JFRudzinski replied:
removed

ka-sarthak commented:
@JFRudzinski Thanks for resolving the previous comments. I added a couple more for the new changes you added. 

To be consistent with other reader tests, can you add the reference JSON for `.raw` under `tests/data`? I added some comments on how I did this for other test files.

JFRudzinski committed:
xray wavelengths to 6 decimal places and add ref

JFRudzinski committed:
reintroduce sample_id parsing

JFRudzinski committed:
standardize raw test with json ref

JFRudzinski committed:
ruff

JFRudzinski commented:
> @JFRudzinski Thanks for resolving the previous comments. I added a couple more for the new changes you added.
> 
> To be consistent with other reader tests, can you add the reference JSON for `.raw` under `tests/data`? I added some comments on how I did this for other test files.

done.

JFRudzinski commented:
@ka-sarthak Thanks again for your review comments. I have now addressed the second round of comments, with only one unresolved discussion (see above).

Please check again at your convenience and let me know if there are further adjustments needed.

ka-sarthak commented:
@JFRudzinski Thanks a lot for making the changes. The PR looks great! I will go ahead and merge it.
